Source/PhysicalParticleContainer.cpp

```diff
@@ -1428,6 +1438,103 @@ PhysicalParticleContainer::Evolve (int lev,
+            long np_split = 6;
+#endif
+            for (int ishift = -1; ishift < 2; ishift +=2 ){
+#if (AMREX_SPACEDIM==2)
+                // Add one particle with offset in x
+                psplit_x.push_back( xp[i] + ishift*dx[0]/2 );
+                psplit_y.push_back( 0. );
+                psplit_z.push_back( zp[i] );
+                psplit_ux.push_back( uxp[i] );
+                psplit_uy.push_back( 0. );
+                psplit_uz.push_back( uzp[i] );
+                psplit_w.push_back( wp[i]/np_split );
+                // Add one particle with offset in z
+                psplit_x.push_back( xp[i] );
+                psplit_y.push_back( 0. );
+                psplit_z.push_back( zp[i] + ishift*dx[2]/2 );
+                psplit_ux.push_back( uxp[i] );
+                psplit_uy.push_back( 0. );
+                psplit_uz.push_back( uzp[i] );
+                psplit_w.push_back( wp[i]/np_split );
+#elif (AMREX_SPACEDIM==3)
+                // Add one particle with offset in x
+                psplit_x.push_back( xp[i] + ishift*dx[0]/2 );
+                psplit_y.push_back( yp[i] );
+                psplit_z.push_back( zp[i] );
+                psplit_ux.push_back( uxp[i] );
+                psplit_uy.push_back( uyp[i] );
+                psplit_uz.push_back( uzp[i] );
+                psplit_w.push_back( wp[i]/np_split );
+                // Add one particle with offset in y
+                psplit_x.push_back( xp[i] );
+                psplit_y.push_back( yp[i] + ishift*dx[1]/2);
+                psplit_z.push_back( zp[i] );
+                psplit_ux.push_back( uxp[i] );
+                psplit_uy.push_back( uyp[i] );
+                psplit_uz.push_back( uzp[i] );
+                psplit_w.push_back( wp[i]/np_split );
+                // Add one particle with offset in x
+                psplit_x.push_back( xp[i] );
+                psplit_y.push_back( yp[i] );
+                psplit_z.push_back( zp[i] + ishift*dx[2]/2 );
+                psplit_ux.push_back( uxp[i] );
+                psplit_uy.push_back( uyp[i] );
+                psplit_uz.push_back( uzp[i] );
+                psplit_w.push_back( wp[i]/np_split );
+#endif
+            }
+            Print()<<"AddNParticles start\n";
+            pctmp_split.AddNParticles(lev, 
+                                       np_split,
+                                       psplit_x.dataPtr(),
+                                       psplit_y.dataPtr(),
+                                       psplit_z.dataPtr(),
+                                       psplit_ux.dataPtr(),
+                                       psplit_uy.dataPtr(),
+                                       psplit_uz.dataPtr(),
+                                       1,
+                                       psplit_w.dataPtr(),
+                                       1);
+            Print()<<"AddNParticles end\n";
+            p.m_idata.id = -p.m_idata.id; // invalidate the particle
+        }
+    }
```

Comment:
I don't think the way this is currently organized will work. I think you need to loop over all the tiles, add the split particles to vectors as you are doing, then use `AddNParticles` outside the loop over `pti`, which will call `Redistribute` on `pctmp` under the hood. After that, also outside the loop, the particles can be added to the regular container without calling `Redistribute`. 

Alternatively, you could add the particles directly to `pctmp` in the main loop using `pctmp.GetParticles(lev)`, getting the tile in `pctmp` corresponding to `pti`, then adding particles to that. You'd still need to call Redistribute and copy the particles to the main container outside this loop.

Reply:
Thank you, I think I will go with the first solution.
As for the copy to the main container, do you want to have a function in AMReX for that or should I do something custom in WarpX?

Reply:
I'll add a function to AMReX now.